### PR TITLE
fix: Improve FindNearbyPlayersSystem performance

### PR DESCRIPTION
**Pull request by askneller**
Addresses #19. Refactoring of the ```FindNearbyPlayersSystem```

Also adds ```AttackInProximityComponent``` that was in some of the hostile critter prefabs. And also adjusts ```CheckAttackStopAction``` to consider ```AttackInProximityComponent```.

I noted a good performance improvement on my system.

Here is the before debug output:
before

And the after:
after

**How to Test**

I did the above tests using the Core Gameplay module with WildAnimals enabled. To make testing easier I changed the reference to the deer prefab in ```WildAnimalsSpawnSystem``` (line 80) to ```WildAnimals:HostileDeer``` so that it spawns lots of entities that will look for players.

If you want to recreate on your system the seed for the above image is ```0pSFAX7zlh7pIlNnLj0DxOfJnzoUsoYc```

### CODE REVIEW

File: src/main/java/org/terasology/behaviors/system/FindNearbyPlayersSystem.java [hunk 1]
```diff
@@ -46,47 +51,48 @@
...
-                        minDistanceFromCharacter = locationComponent.getWorldPosition().distanceSquared(actorPosition);
-                    } else {
-                        if (locationComponent.getWorldPosition().distanceSquared(actorPosition) < minDistanceFromCharacter) {
-                            closestCharacter = character;
-                            minDistanceFromCharacter = locationComponent.getWorldPosition().distanceSquared(actorPosition);
-                        }
-                    }
-                    charactersWithinRange.add(character);
-                }
+            List<Vector3f> inRange = locationSet.stream()
```
Comment (skaldarnar) [inline src/main/java/org/terasology/behaviors/system/FindNearbyPlayersSystem.java:78]: me likes! 👍

File: src/main/java/org/terasology/behaviors/system/FindNearbyPlayersSystem.java [hunk 2]
```diff
@@ -46,47 +51,48 @@
...
+                    .filter(loc -> loc.distanceSquared(actorPosition) <= maxDistanceSquared)
+                    .collect(Collectors.toList());
+            if (inRange.isEmpty()) {
+                findNearbyPlayersComponent.charactersWithinRange = Collections.emptyList();
+                findNearbyPlayersComponent.closestCharacter = EntityRef.NULL;
+                entity.saveComponent(findNearbyPlayersComponent);
+                continue;
             }
 
+            inRange.sort(Comparator.comparingDouble(v3f -> v3f.distanceSquared(actorPosition)));
```
Comment (skaldarnar) [inline src/main/java/org/terasology/behaviors/system/FindNearbyPlayersSystem.java:88]: Why not sort directly when turning the stream into a list (after filtering)? If the list is empty sorting should be cheap, and if it is not we're reaching this line anyways...

File: src/main/java/org/terasology/behaviors/system/FindNearbyPlayersSystem.java [hunk 3]
```diff
@@ -46,47 +51,48 @@
...
+            List<EntityRef> charactersWithinRange = new ArrayList<>(inRange.size());
+            inRange.forEach(v3f -> charactersWithinRange.add(clientLocationMap.get(v3f)));
```
Comment (skaldarnar) [inline src/main/java/org/terasology/behaviors/system/FindNearbyPlayersSystem.java:90]: Why not use Streams-API here, too? Isn't this the same as:

```suggestion
            List<EntityRef> charactersWithinRange = 
            	inRange.stream().map(clientLocationMap::get).collect(Collectors.toList());;
```

File: src/main/java/org/terasology/behaviors/components/AttackInProximityComponent.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright 2017 MovingBlocks
```
Comment (skaldarnar) [inline src/main/java/org/terasology/behaviors/components/AttackInProximityComponent.java:2]: This is so ... last-last-last year 😋

File: src/main/java/org/terasology/behaviors/components/AttackInProximityComponent.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+ * Represents an NPC that will attack players when they enter {@code maxDistance} range.
+ */
+public class AttackInProximityComponent implements Component {
+    // Maximum distance from instigator after which the animal will stop chasing to attack
+    public float maxDistance = 10f;
+    // Speed factor by which attack speed increases
+    public float speedMultiplier = 1.2f;
+    public EntityRef instigator;
+    public long timeWhenHit;
+}
```
Comment (skaldarnar) [inline src/main/java/org/terasology/behaviors/components/AttackInProximityComponent.java:31]: Please end the file with a new line.

```suggestion
}

```

File: src/main/java/org/terasology/behaviors/system/FindNearbyPlayersSystem.java (hunk 1 above)
Comment (askneller) [reply src/main/java/org/terasology/behaviors/system/FindNearbyPlayersSystem.java:78]: :)

File: src/main/java/org/terasology/behaviors/components/AttackInProximityComponent.java (hunk 4 above)
Comment (askneller) [reply src/main/java/org/terasology/behaviors/components/AttackInProximityComponent.java]: Whoops, updated

**Comment by askneller** [comment]
Thanks for the review @skaldarnar, I've implemented those changes.

File: src/main/java/org/terasology/behaviors/system/FindNearbyPlayersSystem.java [hunk 6]
```diff
@@ -46,47 +51,48 @@
...
+                    .collect(Collectors.toList());
+            if (inRange.isEmpty()) {
+                findNearbyPlayersComponent.charactersWithinRange = Collections.emptyList();
+                findNearbyPlayersComponent.closestCharacter = EntityRef.NULL;
+                entity.saveComponent(findNearbyPlayersComponent);
+                continue;
             }
 
+            List<EntityRef> charactersWithinRange = 
+            	inRange.stream().map(clientLocationMap::get).collect(Collectors.toList());;
```
Comment (skaldarnar) [inline src/main/java/org/terasology/behaviors/system/FindNearbyPlayersSystem.java:90]:
```suggestion
            	inRange.stream().map(clientLocationMap::get).collect(Collectors.toList());
```

**Pull request merged**